upup/pkg/fi/cloudup/gce/tpm/gcetpmverifier/tpmverifier.go

```diff
@@ -0,0 +1,247 @@
+limitations under the License.
+*/
+
+package gcetpmverifier
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"math"
+	"net/http"
+	"strings"
+	"time"
+
+	compute "google.golang.org/api/compute/v1"
+	"google.golang.org/api/googleapi"
+	"k8s.io/kops/pkg/bootstrap"
+	"k8s.io/kops/pkg/nodeidentity/gce"
+	"k8s.io/kops/upup/pkg/fi"
+	gcetpm "k8s.io/kops/upup/pkg/fi/cloudup/gce/tpm"
+)
+
+type tpmVerifier struct {
+	opt gcetpm.TPMVerifierOptions
+
+	computeClient *compute.Service
+}
+
+// NewTPMVerifier constructs a new TPM verifier for GCE.
+func NewTPMVerifier(opt *gcetpm.TPMVerifierOptions) (bootstrap.Verifier, error) {
+	ctx := context.Background()
+
+	computeClient, err := compute.NewService(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error building compute API client: %w", err)
+	}
+
+	return &tpmVerifier{
+		opt:           *opt,
+		computeClient: computeClient,
+	}, nil
+}
+
+var _ bootstrap.Verifier = &tpmVerifier{}
+
+func (v *tpmVerifier) VerifyToken(ctx context.Context, authToken string, body []byte) (*bootstrap.VerifyResult, error) {
+	if !strings.HasPrefix(authToken, gcetpm.GCETPMAuthenticationTokenPrefix) {
+		return nil, fmt.Errorf("incorrect authorization type")
+	}
+	authToken = strings.TrimPrefix(authToken, gcetpm.GCETPMAuthenticationTokenPrefix)
+
+	tokenBytes, err := base64.StdEncoding.DecodeString(authToken)
+	if err != nil {
+		return nil, fmt.Errorf("decoding authorization token: %w", err)
+	}
+
+	token := &gcetpm.AuthToken{}
+	err = json.Unmarshal(tokenBytes, token)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling authorization token: %w", err)
+	}
+
+	// Guard against replay attacks
+	if token.Data.Audience != gcetpm.AudienceNodeAuthentication {
+		return nil, fmt.Errorf("incorrect Audience")
+	}
+	timeSkew := math.Abs(time.Since(time.Unix(token.Data.Timestamp, 0)).Seconds())
+	if timeSkew > float64(v.opt.MaxTimeSkew) {
+		return nil, fmt.Errorf("incorrect Timestamp %v", token.Data.Timestamp)
+	}
+
+	// Verify the token has signed the body content.
+	requestHash := sha256.Sum256(body)
+	if !bytes.Equal(requestHash[:], token.Data.RequestHash) {
+		return nil, fmt.Errorf("incorrect RequestHash")
+	}
+
+	// Some basic validation to avoid requesting invalid instances.
+	if token.Data.GCPProjectID == "" {
+		return nil, fmt.Errorf("gcpProjectID is required")
+	}
+	if token.Data.Zone == "" {
+		return nil, fmt.Errorf("zone is required")
+	}
+	if token.Data.Instance == "" {
+		return nil, fmt.Errorf("instance is required")
+	}
+
+	// Verify node is in our cluster
+	if token.Data.GCPProjectID != v.opt.ProjectID {
+		return nil, fmt.Errorf("projectID does not match expected: got %q, want %q", token.Data.GCPProjectID, v.opt.ProjectID)
+	}
+
+	instance, err := v.computeClient.Instances.Get(token.Data.GCPProjectID, token.Data.Zone, token.Data.Instance).Do()
```

Comment:
Consider the possibility of the client including '/' or other problematic character in the zone or instance field. Could that cause the computClient implementation to misbehave?

Reply:
Such a good point.  We shouldn't be trusting this data until we've checked the signature.  I had a look and the google SDK does do the right thing in terms of escaping (I'd feel happier if it was using proto, but the escaping logic looks pretty sound).  Adding a comment though to remind ourselves to be careful!